### ISSUE TITLE
(System.IO.Ports.SerialPort) PinChanged event never triggers on CTS state change in Linux 

### ISSUE THREAD

**Issue by liberza** [dotnet/corefx#39814]
I ran into trouble with the PinChanged event, where it does not get triggered when the CTS line changes state. Manually checking CTS seems to work but the event never gets triggered. I didn't see any open issues, and it's not listed in https://github.com/dotnet/corefx/issues/33146, but I noticed that this test is also marked as KnownFailure: https://github.com/dotnet/corefx/blob/master/src/System.IO.Ports/tests/SerialPort/PinChangedEvent.cs#L23

**Comment by liberza** [comment]
Apparently the PinChanged event in SerialStream is never called on Linux:
https://github.com/dotnet/corefx/blob/master/src/System.IO.Ports/src/System/IO/Ports/SerialStream.cs#L24

**Comment by stephentoub** [comment]
cc: @krwq

**Comment by krwq** [comment]
@liberza that's possible I missed this limitation - I wrote down group of failures I observed when working on async implementation and possibly this got covered by different issue.

Possibly we could use https://stackoverflow.com/a/8952165 (ioctl + TIOCMIWAIT) or just select/poll/epoll.

Would you be interested in sending PR with the fix?

**Comment by liberza** [comment]
For now we're going to give SerialPortStream a shot and see if it fits our needs.

If full Linux support for 3.0 is not planned, please consider updating [this page](https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0#serialport-for-linux) to say ".NET Core 3.0 has limited support of System.IO.Ports.SerialPort on Linux". My takeaway from that doc was that all features in SerialPort were available in Linux.

**Comment by krwq** [comment]
@liberza created PR against docs to clarify this. Please let me know if there are any specific things which would you like to see sooner than later as it may help us with prioritization.

**Comment by krwq** [comment]
I've just ordered 2 USB/U(S)ART with CTS/RTS in case I will have time to pick it up (can't promise I will though)

**Comment by liberza** [comment]
@krwq as for specific things, it's mainly just the event for PinChange events, specifically the CTS case. In our use-case, the CTS line indicates the presence of a physical part of the serial circuit. DTR and RTS also need to be settable from the sender-side, but I haven't tested those under Linux yet using SerialPort. They did work under Windows though.

**Comment by krwq** [comment]
@liberza I **think** setting pins work (at least I recall some code out there) but I do not recall code for events (and you've also checked yourself it's not being called)

**Comment by JeremyKuhne** [comment]
@krwq Is this a 3.0 issue?

**Comment by krwq** [comment]
@JeremyKuhne considering the docs are updated and the plan was to add basic support I have mixed feelings (I vote to move this to 5.0 although I already bought and received hardware to test and fix this scenario)

**Comment by krwq** [comment]
@liberza if you can, please test the PR locally (build corefx, you should see nuget package in the output which you can consume with local nuget package source)

**Comment by liberza** [comment]
I can try and test this, I'm using dotnet publish on Win10 to create a self-contained deployment for Linux though. I haven't built corefx before, does that make things more complicated?

**Comment by krwq** [comment]
@liberza when you build from sources you can use docker to cross build for linux. I.e. something like:

here is how example instructions could look like
(assuming you already cloned corefx repo)

make sure to replace all of the placeholders, I haven't tested it so might need adjustments

```batch
set DOCKER_IMAGE=<corefx docker image, i.e.: "mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-19.04-helix-amd64-09ca40b-20190508143257", you can also find a better suiting one on dockerhub website>
set COREFX_PATH=<your corefx path>
set COREFX_REMOTE_PATH=<corefx path inside docker container, you can start a container without those args first and inspect which path works for you>
cd "%COREFX_PATH%"
docker run -t 
docker run -it --rm --privileged --network host -v "%COREFX_PATH%":"%COREFX_REMOTE_PATH%" -w "%COREFX_REMOTE_PATH%"
```

And then in the container

```shell
# basically follow regular corefx build instructions: https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/developer-guide.md
sudo apt-get -qy update
sudo apt-get -qy install \
    git cmake make libc6-dev libssl-dev libkrb5-dev libcurl4-openssl-dev zlib1g-dev wget
sudo locale-gen en_US.UTF-8
export LC_ALL=en_US.UTF-8
export LANG=en_US.UTF-8

# cd into your remote corefx path
# ./build.sh # this is just commented out in case you don't read above comment
```

after you build assuming you passed `-v` correctly you can exit container and will see your binaries in your windows and then you can copy them to your desired OS (which should be similar to your docker image)

**Comment by liberza** [comment]
Unfortunately it looks like that didn't fix things on my end. Just to make sure I'm using the build correctly, here's what I did. Maybe I'm missing something.

1. Built your corefx branch in WSL (Ubuntu 18.04)
2. Refactored my code to use System.IO.Ports instead of the SerialPortStream library. This was straightforward as I had used System.IO.Ports originally.
3. Published to my remote CentOS 7 box as I've been doing normally.
4. Replaced the published System.IO.Ports.(dll | so | a) with the newly-built version.
5. Dropped a breakpoint into my PinChanged callback, started remote debugging.
6. Breakpoint never gets triggered when changing CTS line.

**Comment by krwq** [comment]
@liberza I'm unsure the binaries built on Ubuntu 18.04 can be moved to CentOS (I suspect they can't)

Did you try the sample code I wrote in here https://github.com/dotnet/corefx/pull/40383#issue-308248089 ? This has worked correctly for me.

Can you check under the debugger if the binaries you have replaced are being used by your exe?

On my ubuntu 18.10 the nuget packages are being built to

```
./artifacts/bin/runtime/netcoreapp-Linux-Debug-x64/System.IO.Ports.pdb
./artifacts/bin/runtime/netcoreapp-Linux-Debug-x64/System.IO.Ports.Native.so
./artifacts/bin/runtime/netcoreapp-Linux-Debug-x64/System.IO.Ports.Native.a
./artifacts/bin/runtime/netcoreapp-Linux-Debug-x64/System.IO.Ports.dll
```

where are you putting binaries to?

**Comment by wfurt** [comment]
moving to different LInux should generally work as long as it has older (g)libc. Other catch is that Ubuntu uses symbol versioning for some libraries. But that would fail to load `.so`  with loader error.

**Comment by liberza** [comment]
> Did you try the sample code I wrote in here #40383 (comment) ? This has worked correctly for me.

I have not yet, but mine is similar enough that I figure it's best to verify the stuff with the build first.

> Can you check under the debugger if the binaries you have replaced are being used by your exe?

I didn't check in the debugger, but I'm pretty confident it's using the "new" ones that I copied over since it's a self-contained deploy and I deleted the old ones.

> What you described sounds like your exe still uses old binaries

I agree, it's the same behavior. I'll try copying them from the `./artifacts/bin/runtime` dir next time, maybe I just grabbed old binaries that were in the build folder. It'll probably be around next week when I get to try again.
